### ISSUE TITLE
Document how to create a Sponge permission using only Forge

### ISSUE THREAD

**Issue by MoeBoy76** [SpongePowered/SpongeDocs#361]
Not sure where it should go so i'm putting it here for other brains to work on. The very simple way to create and check for a permission in a Forge mod is to use:

``` java
if(player.canCommandSenderUseCommand(4, "hi.i.am.a.permission")) {
    //Do stuff
}
```

what this does in Forge is checks for OP (perm level 4) or on a SpongeForge server with a permissions plugin it checks if the player has "hi.i.am.permission". A better example is the Pixelmon NPCEditor, on right-clicking the above check is done for the permission "pixelmon.npceditor.use".

This is based on Noppes asking in the Pixelmon dev skype how to convert from his atrocious bukkit permission code to support Sponge and i'm sure others want to know how as well.

**Comment by RobertHerhold** [comment]
I'm not sure what you're saying... are you proposing a method to the API? If so, it belongs [here](https://github.com/SpongePowered/SpongeAPI/issues)

**Comment by MoeBoy76** [comment]
no, i mean i think it should be noted somewhere that to add permissions to a Forge mod without having a dependency on SpongeAPI you use `ICommandSender.canCommandSenderUseCommand(int permLevel, String commandName)`

**Comment by ryantheleach** [comment]
Are you sure? I can't find that method in Forge... https://github.com/MinecraftForge/MinecraftForge/search?utf8=%E2%9C%93&q=canCommandSenderUseCommand

**Comment by MoeBoy76** [comment]
`net.minecraft.command.ICommandSender.canCommandSenderUseCommand(int permLevel, String commandName);` if you want me to be 100% specific, usually used on either an instance of `net.minecraft.entity.player.EntityPlayerMP` or simply a `net.minecraft.entity.player.EntityPlayer`

the point i'm trying to make is using this method is a lot quicker and easier that adding the dependency for SpongeAPI, casting to a `org.spongepowered.api.service.permission.Subject` and using `hasPermission(String permission)` for something that is only used server side

**Comment by simon816** [comment]
`canCommandSenderUseCommand` takes an integer permission level and a String command. It's not a `String permission` and is not going to work to test sponge-style permissions.
in Vanilla, permission is defined as an integer, in sponge it's a node tree and you can't just put that node as a command.

Are you asking to allow this to be possible?

**Comment by MoeBoy76** [comment]
time for a story, Pixelmon only allows OPs to edit our PixelmonSpawner block by using the above method, when testing on one of our public test servers i found that my moderators couldn't use them (when Opped) so i decided to add the permission for an empty string, low and behold it worked, so i changed the string to "pixelmon.spawner.use" and they then needed that as the permission.

tl;dr Sponge creates a permission based on the string cos it is a command

**Comment by clienthax** [comment]
Basically Sponge modifies the method to call onwards to the player perms in MixinICommandSender

**Comment by MoeBoy76** [comment]
Edited the title to better reflect the content and reduce confusion

**Comment by ryantheleach** [comment]
does it prefix the commands at all? or is it based on the commands alias?

**Comment by MoeBoy76** [comment]
it doesn't prefix it at all, possibly cos the "command" in the given string is not registered anywhere so there is no way to really detect the modid, here's a new example

```java
public Block AwesomeBlock extends Block {
    @Override
    public boolean onBlockActivated(World world, BlockPos pos, IBlockState state, EntityPlayer player, EnumFacing side, float hitX, float hitY, float hitZ) {
        if (player instanceof EntityPlayerMP) {
            if(player.canCommandSenderUseCommand(4, ExampleMod.MODID + ".awesomeBlock.interact")) {
                //Do cool stuff
                return true;
            }
        }
        return false;
    }
}
```
if you were checking if a player can interact with your mod block then SpongeAPI would change it to
```java
player.hasPermission("examplemod.awesomeBlock.interact");
```

**Comment by ryantheleach** [comment]
I get that, but I was thinking for the vanilla minecraft commands.

**Comment by MoeBoy76** [comment]
this doesn't have anything to do with commands at all, except for the name of the method in Forge/Vanilla

i opened this issue with the intention of documenting that Forge Mods can use the method ```java ICommandSender.canCommandSenderUseCommand(int permLevel, String commandName)``` to test for OP when the server uses just Forge or to check a permission if the server is using SpongeForge

**Comment by ryantheleach** [comment]
This would make a good addition, bar adding docs regarding the new Forge Permission API.

**Comment by MoeBoy76** [comment]
Fixed by #621
